source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,871 @@
+drivers MUST set the connection’s read timeout to the remaining ``timeoutMS``
+value. The timeout MUST apply to the aggregate of all reads done to receive a
+server response, not to individual reads. If any read or write calls on the
+socket fail with a timeout, drivers MUST transform the error into the new
+timeout exception as described in the `Error Transformations`_ section.
+
+If ``timeoutMS`` is not set and support for ``socketTimeoutMS`` has not been
+removed, drivers MUST honor ``socketTimeoutMS`` as the timeout for socket
+reads and writes.
+
+See `maxTimeMS accounts for server RTT`_.
+
+Batching
+~~~~~~~~
+
+If an operation must be sent to the server in multiple batches (e.g.
+``collection.bulkWrite()``), the ``timeoutMS`` option MUST apply to the
+entire operation, not to each individual batch.
+
+Retryability
+~~~~~~~~~~~~
+
+If an operation requires a retry per the retryable reads or writes
+specifications and ``timeoutMS`` is set to a non-zero value, drivers MUST
+retry operations as many times as possible before the timeout expires or a
+retry attempt returns a non-retryable error. Once the timeout expires, a
+timeout error MUST be raised. If ``timeoutMS`` is unset or is explicitly set
+to zero, drivers MUST retry operations as many times as needed until a retry
+attempt returns a non-retryable error. However, if ``timeoutMS`` is unset but
+the ``socketTimeoutMS`` option is set, drivers MUST stop retrying after
+encountering two socket timeout errors, either consecutive or
+non-consecutive. In this case, the second socket timeout error MUST be
+propagated to the application.
+
+See `Retryable reads/writes behavior`_ and `Why don’t drivers use
+backoff/jitter between retry attempts?`_.
+
+Client Side Encryption
+~~~~~~~~~~~~~~~~~~~~~~
+
+If client-side encryption is enabled, the remaining ``timeoutMS`` value MUST
+be used as the ``timeoutMS`` when executing ``listCollections`` commands to
+retrieve collection schemas, ``find`` commands to get data from the key
+vault, and any commands against mongocryptd. It MUST also be used as the
+request timeout for HTTP requests against KMS servers to decrypt data keys.
+When sending a command to mongocryptd, drivers MUST NOT append a
+``maxTimeMS`` field. This is to ensure that a ``maxTimeMS`` field can be
+safely appended to the command after it has been marked by mongocryptd and
+encrypted by libmongocrypt. To determine whether or not the server is a
+mongocryptd, drivers MUST check that the ``iscryptd`` field in the server's
+description is ``true``.
+
+See `maxTimeMS is not added for mongocryptd`_.
+
+Background Connection Pooling
+-----------------------------
+
+Connections created as part of a connection pool’s ``minPoolSize``
+maintenance routine MUST use ``connectTimeoutMS`` as the timeout for
+connection establishment. After the connection is established,
+``connectTimeoutMS`` MUST be used as the ``timeoutMS`` value for all commands
```

Comment:
I'm concerned that `connectTimeoutMS` could be configured as too short, particularly for heavyweight authentication mechanisms like LDAP or Kerberos.

Is it necessary to have a timeout for background connection establishment at all?  What's the benefit?

Reply:
The default `connectTimeoutMS` is 10 seconds, as defined in the URI options spec. IIRC I had a quick conversation with the server security team and they agreed that 10 seconds should be enough time for any one auth command to succeed, regardless of the auth mechanism. FWIW, the motivation to apply `connectTimeoutMS` rather than `timeoutMS` here came from SPEC-1564, but we revisit this if you feel that `timeoutMS` would be more suitable.

As for the benefit, I think it's more to prevent backwards breaking changes. We're already applying `socketTimeoutMS` to auth commands, but that's now a deprecated option and SPEC-1564 outlines some reasoning for making this `connectTimeoutMS` instead.

Reply:
Ten seconds is the default, and that does seem sufficient, but what if the application configures it to be much shorter?  Say, long enough to connect to mongo server on a fast connection but too short to finish a Kerberos auth request to a perhaps more remote KDC.

Reply:
Do you think we should close SPEC-1564 with the rationale that it could break existing applications? If so, we're currently applying `socketTimeoutMS` to auth/handshake commands for connections created in the background, so we can probably change that to apply `timeoutMS` instead.

Reply:
As discussed offline, this section has been changed to use `timeoutMS` instead of `connectTimeoutMS` for handshake/auth commands. I also added a sentence to say that we should continue to honor `socketTimeoutMS` for these commands if `timeoutMS` is not set. This [commit range](https://github.com/mongodb/specifications/pull/881/files/f4e70a8f8c76d414262f3b82f0e3635810cf50f6..31c3452bb79b5bdca8c14f0e4cac4322cd8dde79) contains the relevant changes.